File: test/test_init.py
import stat

import pytest

from helpers import Git


def test_hook_installed(git: Git) -> None:
    git.init_repo()
    hook_path = git.path / ".git" / "hooks" / "post-commit"
    assert hook_path.exists()
    assert hook_path.stat().mode & stat.S_IXUSR
    assert hook_path.stat().mode & stat.S_IXGRP
    assert hook_path.stat().mode & stat.S_IXOTH


def test_alias_installed(git: Git) -> None:
    git.init_repo()
    assert (
        git.run("smartlog")
        == """\
@ f777ecc9 (master) create initial.txt
"""
    )
    assert (
        git.run("sl")
        == """\
@ f777ecc9 (master) create initial.txt
"""
    )


def test_old_git_version_warning(git: Git) -> None:
    git.init_repo()
    if git.get_version() >= (2, 29, 0):
        pytest.skip("Requires Git version earlier than v2.29")
    assert "requires Git v2.29" in git.run("branchless", ["init"]).replace("\n", " ")

File: tests/test_init.rs
use anyhow::Context;
use branchless::util::GitVersion;

#[cfg(test)]
#[test]
fn test_hook_installed() -> anyhow::Result<()> {
    branchless::testing::with_git(|git| {
        git.init_repo()?;
        let hook_path = git.repo_path.join(".git").join("hooks").join("post-commit");
        assert!(hook_path.exists());

        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let metadata = std::fs::metadata(&hook_path)
                .with_context(|| format!("Reading hook permissions for {:?}", &hook_path))?;
            let mode = metadata.permissions().mode();
            assert!(mode & 0o111 == 0o111);
        }

        Ok(())
    })
}

#[test]
fn test_alias_installed() -> anyhow::Result<()> {
    branchless::testing::with_git(|git| {
        git.init_repo()?;

        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @r###"
@ f777ecc9 (master) create initial.txt
"###);
        }

        {
            let (stdout, _stderr) = git.run(&["sl"])?;
            insta::assert_snapshot!(stdout, @r###"
@ f777ecc9 (master) create initial.txt
"###);
        }

        Ok(())
    })
}

#[test]
fn test_old_git_version_warning() -> anyhow::Result<()> {
    branchless::testing::with_git(|git| {
        git.init_repo()?;
        let version = git.get_version()?;
        if version < GitVersion(2, 29, 0) {
            let (stdout, _stderr) = git.run(&["branchless", "init"])?;
            assert!(stdout.replace("\n", " ").contains("requires Git v2.29"));
        }

        Ok(())
    })
}
